### PR TITLE
plutus-ledger: use de-bruijn indices for Scripts

### PR DESCRIPTION
**Pull request by michaelpj**
Fixes SCP-616

This corresponds to using de-bruijn indices on-chain, since `Script`s
are our representation of "a PLC program in a transaction".

Note, this currently has an `error` in it. I'm trying to decide whether it's evil or whether it should be turned into a proper `throw` of some kind of "impossible" exception. It can only happen if the script does not typecheck, which shouldn't be possible at this stage.

### CODE REVIEW

File: plutus-core/src/Language/PlutusCore/DeBruijn/Internal.hs [hunk 1]
```diff
@@ -39,20 +40,25 @@ import           Data.Typeable
 
 import           Numeric.Natural
 
+import           Control.DeepSeq            (NFData)
 import           GHC.Generics
 
 -- | A relative index used for de Bruijn identifiers.
 newtype Index = Index Natural
```
Comment (kwxm) [inline plutus-core/src/Language/PlutusCore/DeBruijn/Internal.hs:47]: What's our policy on using `Natural`?  As far as I can see, the only place where we use these in `plutus-core` is here and in PLC version numbers.  It seems a bit odd that we use some very complicated type-level stuff to enforce invariants on data, but almost never use Natural in places where negative numbers should never occur.

File: plutus-ledger/src/Ledger/Scripts.hs [hunk 2]
```diff
@@ -137,7 +137,9 @@ fromCompiledCode :: CompiledCode PLC.DefaultUni a -> Script
 fromCompiledCode = fromPlc . getPlc
 
 fromPlc :: UPLC.Program PLC.Name PLC.DefaultUni () -> Script
-fromPlc = Script
+fromPlc p = case UPLC.deBruijnProgram p of
+    Right p' -> Script p'
+    Left _   -> Haskell.error "Debruijn failed"
```
Comment (kwxm) [inline plutus-ledger/src/Ledger/Scripts.hs:142]: I feel that I should quibble about the capitalisation in the error message.

**Review by kwxm — COMMENTED** [review]
I had a very quick look at this.  It looks basically OK, but I'm worried about serialisation.  The `DeBruijn` name type includes textual names, and it looks as if these are still serialised along with the indices.  I think we really don't want that.  Including the names increases the size of the serialised scripts signficantly: see https://github.com/input-output-hk/plutus/blob/master/notes/plutus-core/Merklisation/Erasure.md#results.  The column headed "Untyped, de Bruijn` actually refers to serialised ASTs that only include the numbers from the de Bruijn indices.  Unfortunately the table doesn't contains figures for de Bruijn indices including textual names, but the columns for other versions of variable names show that they take up  a lot of room.  I think that on the chain "names" should just be integers.

We should also look at how using de Bruijn indices affects the `flat` sizes,

**Comment by michaelpj** [comment]
> The DeBruijn name type includes textual names, and it looks as if these are still serialised along with the indices. I think we really don't want that.

Yes, I agree with that. The annoyance is that we also use this for testing, and there it's quite useful to have the names along with the indices. Maybe we need two ways to do it... I'll have a think.

**Comment by kwxm** [comment]
I also wonder whether we can use de Bruijn indices/levels directly in the CEK machine.  See https://jira.iohk.io/browse/SCP-1196.

**Comment by kwxm** [comment]
> > The DeBruijn name type includes textual names, and it looks as if these are still serialised along with the indices. I think we really don't want that.
> 
> Yes, I agree with that. The annoyance is that we also use this for testing, and there it's quite useful to have the names along with the indices. Maybe we need two ways to do it... I'll have a think.

Yeah, PLC's incomprehensible enough already; de Bruijn indices aren't going to help.

**Comment by michaelpj** [comment]
> I also wonder whether we can use de Bruijn indices/levels directly in the CEK machine. See https://jira.iohk.io/browse/SCP-1196.

At the moment this will save us `O(program size)` work before starting evaluation, which seems plausibly worth it, but at the moment we're just going to eat it, I guess.

**Comment by michaelpj** [comment]
Now switches to a name-string-less debruijn index form for `Script`s (and hence serialization).

**Comment by michaelpj** [comment]
(I copied over some more functions for UPLC that I ended up not needing, but I'll need some of them in a subsequent PR, so I just left them in for now.)

**Comment by bezirg** [comment]
Up until now, looks good.

**Pull request merged**